development_bank_wales/analysis/Initial Upgradability Analysis.py

```diff
@@ -0,0 +1,634 @@
+        wales_df,
+        "PROPERTY_TYPE",
+        rec,
+        plotting_colors="copper",
+        x_tick_rotation=45,
+        figsize=(7, 5),
+        fig_save_path=fig_output_path,
+    )
+    easy_plotting.plot_subcats_by_other_subcats(
+        wales_df,
+        "BUILT_FORM",
+        rec,
+        plotting_colors="copper",
+        x_tick_rotation=45,
+        figsize=(7, 5),
+        fig_save_path=fig_output_path,
+    )
+    easy_plotting.plot_subcats_by_other_subcats(
+        wales_df,
+        "CONSTRUCTION_AGE_BAND",
+        rec,
+        plotting_colors="copper",
+        x_tick_rotation=45,
+        figsize=(7, 5),
+        fig_save_path=fig_output_path,
+    )
+    easy_plotting.plot_subcats_by_other_subcats(
+        wales_df,
+        "CURRENT_ENERGY_RATING",
+        rec,
+        plotting_colors="copper",
+        x_tick_rotation=45,
+        figsize=(7, 5),
+        fig_save_path=fig_output_path,
+    )
+    easy_plotting.plot_subcats_by_other_subcats(
+        wales_df,
+        "LOCAL_AUTHORITY_LABEL",
+        rec,
+        plotting_colors="copper",
+        x_tick_rotation=45,
+        figsize=(7, 5),
+        fig_save_path=fig_output_path,
+    )
+    easy_plotting.plot_subcats_by_other_subcats(
+        wales_df,
+        "TENURE",
+        rec,
+        plotting_colors="copper",
+        x_tick_rotation=45,
+        figsize=(7, 5),
+        fig_save_path=fig_output_path,
+    )
+
+
+# %% [markdown]
+# ### Recommendations over Time<a id='recs_over_time'></a>
+# [[back to top]](#top)
+#
+# Looking at properties with multiple records over time, we can estimate which upgrades are often carried out and which recommendations are implemented. Note that this only an approximation as we do not hold information for all properties, and only one record for many.
+
+# %%
+wales_df = wales_df.loc[wales_df["TENURE"] == "owner-occupied"]
+
+
+latest_wales = epc_data.filter_by_year(
+    wales_df.loc[wales_df["N_SAME_UPRN_ENTRIES"] > 1],
+    None,
+    selection="latest entry",
+)
+first_wales = epc_data.filter_by_year(
+    wales_df.loc[wales_df["N_SAME_UPRN_ENTRIES"] > 1],
+    None,
+    selection="first entry",
+)
+
+# %%
+for rec in unique_recs:
+
+    first_wales_with_rec = recommendations.check_for_implemented_rec(
+        rec, first_wales, latest_wales, identifier="UPRN", keep="first"
+    )
+    print(round(first_wales_with_rec[rec].value_counts(normalize=True), 2))
+    print()
+
+
+# %%
+@interact(rec=recs, first_wales=fixed(first_wales), latest_wales=fixed(latest_wales))
+def plot_recommendations(rec, first_wales, latest_wales):
+
+    first_wales_with_rec = recommendations.check_for_implemented_rec(
+        rec, first_wales, latest_wales, keep="first", identifier="UPRN"
+    )
+
+    print(rec)
+    print(
+        "IMPLEMENTED_" + rec,
+    )
+
+    # Only look at samples with this recommendation
+    first_wales_with_rec = first_wales_with_rec.loc[first_wales_with_rec[rec]]
```

Comment:
I don't think the `loc` is needed in this specific line of code (and there are other cases like this throughout the code):
```suggestion
    first_wales_with_rec = first_wales_with_rec[first_wales_with_rec[rec]]
```